lib/domain/validation/authenticator_validator.dart

```diff
@@ -0,0 +1,116 @@
+    List<String> authenticatorAllowList,
+  ) {
+    Set<Authenticator> allowedAuthenticators = _allowedAuthenticators(
+      context,
+      authenticatorAllowList,
+    );
+
+    return whereAsync(allowedAuthenticators, (authenticator) async {
+      // Do not display:
+      //  - policy non-compliant authenticators (this includes already registered authenticators)
+      //  - not hardware supported authenticators
+      //  - prefer Biometrics authenticator on Android
+      return authenticator.isSupportedByHardware &&
+          await context.isPolicyCompliant(authenticator.aaid) &&
+          await _filterAndroidFingerprintIfNecessary(context, authenticator);
+    });
+  }
+
+  @override
+  Set<Authenticator> validateForAuthentication(
+    AuthenticatorSelectionContext context,
+    List<String> authenticatorAllowList,
+  ) {
+    return _allowedAuthenticators(
+      context,
+      authenticatorAllowList,
+    ).where((authenticator) {
+      // Do not display:
+      //   - non-registered authenticators
+      //   - not hardware supported authenticators
+      return authenticator.registration
+              .isRegistered(context.account.username) &&
+          authenticator.isSupportedByHardware;
+    }).toSet();
+  }
+
+  Set<Authenticator> _allowedAuthenticators(
+    AuthenticatorSelectionContext context,
+    List<String> authenticatorAllowList,
+  ) {
+    return context.authenticators
+        .where((a) => authenticatorAllowList.contains(a.aaid))
+        .toSet();
+  }
+
+  Future<bool> _filterAndroidFingerprintIfNecessary(
+    AuthenticatorSelectionContext context,
+    Authenticator authenticator,
+  ) async {
+    if (Platform.isIOS) {
+      return Future.value(true);
+    }
+
+    final authenticators = context.authenticators;
+    final isBiometricRegistered = await anyAsync(
+      authenticators,
+      (Authenticator a) => Future.value(
+        a.aaid.isBiometric &&
+            a.registration.isRegistered(context.account.username),
+      ),
+    );
+    final canRegisterBiometric =
+        await anyAsync(authenticators, (Authenticator a) async {
+      return Future.value(
+        a.aaid.isBiometric &&
+            await context.isPolicyCompliant(a.aaid) &&
+            a.isSupportedByHardware,
+      );
+    });
+    final canRegisterFingerprint =
+        await anyAsync(authenticators, (Authenticator a) async {
+      return Future.value(
+        a.aaid.isFingerprint &&
+            await context.isPolicyCompliant(a.aaid) &&
+            a.isSupportedByHardware,
+      );
+    });
+
+    // If biometric can be registered (or is already registered), or if we
+    // cannot register fingerprint, do not propose to register fingerprint
+    // (we favor biometric over fingerprint).
+    if ((canRegisterBiometric ||
+            isBiometricRegistered ||
+            !canRegisterFingerprint) &&
+        authenticator.aaid.isFingerprint) {
+      return Future.value(false);
+    }
```

Comment:
🤔 I'm not entirely sure we need the same logic as in the _Access App_ and I suspect **no**: there we don't filter the _Fingerprint authenticator_ if a fingerprint is already registered and in this case, we filter the Biometric authenticator instead.

I guess this is not relevant here, as we don't want/need to support previous registrations (before the SDK introduced _Biometric authenticator_) so indeed we always prefer _Biometric authenticator_ (and thus filter the Fingerprint one) unless it's not available?!

Reply:
Along with your policy compliance questions I think this is out of the scope of this story. It's about moving the validation to the same place.

On the other hand this is an Android heavy question and if changes are required that will have affect to 4 example apps. We need to clarify it, make a decision and implement it in one of the native example apps. If that will be approved by everyone, the changes can be done in other example apps.
Can you please create a ticket for that? You can provide the best description...thanks!